Well, it works 50% of the times. It extremely flaky. Here are some failures:
```
make test-integration WHAT=./staging/src/k8s.io/kube-apiextensions-server/test/integration/ KUBE_TEST_ARGS="-run TestFinalization"
+++ [0526 21:27:22] Checking etcd is on PATH
/Users/ash2k/gopath/src/k8s.io/kubernetes/third_party/etcd/etcd
+++ [0526 21:27:22] Starting etcd instance
etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/tmp.Ea7v7ceH --listen-client-urls http://127.0.0.1:2379 --debug > "/dev/null" 2>/dev/null
Waiting for etcd to come up.
+++ [0526 21:27:22] On try 1, etcd: : http://127.0.0.1:2379
{"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
+++ [0526 21:27:22] Running integration test cases
Running tests for APIVersion: v1
+++ [0526 21:27:28] Running tests without code coverage
W0526 21:27:31.997487    6573 authentication.go:50] Authentication is disabled
I0526 21:27:32.001158    6573 serve.go:85] Serving securely on 127.0.0.1:50070
I0526 21:27:32.001222    6573 crd_finalizer.go:249] Starting CRDFinalizer
I0526 21:27:32.001237    6573 customresource_discovery_controller.go:152] Starting DiscoveryController
I0526 21:27:32.001244    6573 naming_controller.go:284] Starting NamingConditionController
2017-05-26T21:27:32.005928336+10:00 AUDIT: id="309d0bef-0f49-4917-9baa-5011ca7497b2" ip="127.0.0.1" method="list" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions?resourceVersion=0" response="200"
2017-05-26T21:27:32.006912838+10:00 AUDIT: id="762c4e22-b84f-4861-9db5-a156d9a061c3" ip="127.0.0.1" method="watch" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions?resourceVersion=1&timeoutSeconds=582&watch=true" response="<deferred>"
2017-05-26T21:27:32.036872032+10:00 AUDIT: id="ab0703e5-30d5-48f0-b114-1a4147e38bc2" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/healthz" response="200"
2017-05-26T21:27:32.037569565+10:00 AUDIT: id="5744cc23-af68-4261-9cb1-dc2a4d4e9f4f" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1" response="200"
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"apiextensions.k8s.io/v1alpha1","resources":[{"name":"customresourcedefinitions","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["create","delete","deletecollection","get","list","patch","update","watch"]},{"name":"customresourcedefinitions/status","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["update"]}]}
2017-05-26T21:27:32.039788023+10:00 AUDIT: id="6d17461a-db5d-4303-823b-7ef5c6220092" ip="127.0.0.1" method="create" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions" response="201"
2017-05-26T21:27:32.041934955+10:00 AUDIT: id="5f89c562-5344-4443-8215-835ae71488b3" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="404"
2017-05-26T21:27:32.076868996+10:00 AUDIT: id="121a954c-0d6f-468a-b28d-d0140991fb32" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="404"
2017-05-26T21:27:32.106796844+10:00 AUDIT: id="a696b059-fade-4bf2-a942-e6364b7ed744" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="404"
2017-05-26T21:27:32.107245067+10:00 AUDIT: id="7ac9d6ca-8993-41b8-8f4f-f15b2d5363b9" ip="127.0.0.1" method="update" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions/noxus.mygroup.example.com/status" response="200"
2017-05-26T21:27:32.132624516+10:00 AUDIT: id="082f87ec-7a26-400a-b705-3041f341e5e0" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1" response="200"
2017-05-26T21:27:32.133292298+10:00 AUDIT: id="98b3f276-bdb6-47fd-91da-1861db2be094" ip="127.0.0.1" method="create" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus" response="201"
2017-05-26T21:27:32.134818471+10:00 AUDIT: id="fa6feea7-1b1a-4c6e-92e4-39aa3c54e903" ip="127.0.0.1" method="delete" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus/foo123" response="200"
2017-05-26T21:27:32.139720222+10:00 AUDIT: id="9961b9a9-e86c-4f24-8a38-dfc095e060de" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus/foo123" response="200"
2017-05-26T21:27:32.14065171+10:00 AUDIT: id="d408cd4d-e6ea-402f-a8a4-dfa999ac9a06" ip="127.0.0.1" method="update" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus/foo123" response="409"
2017-05-26T21:27:32.141324285+10:00 AUDIT: id="e9577d53-6830-4045-8d41-f34bcabfed9d" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/mygroup.example.com/v1alpha1/noxus/foo123" response="200"
I0526 21:27:32.142014    6573 crd_finalizer.go:261] Shutting down CRDFinalizer
--- FAIL: TestFinalization (0.60s)
        finalization_test.go:45: get object map[string]interface {}{"metadata":map[string]interface {}{"resourceVersion":"5", "uid":"4f094c89-4206-11e7-bb84-80e65026e166", "creationTimestamp":"2017-05-26T11:27:32Z", "deletionTimestamp":"2017-05-26T11:27:32Z", "finalizers":[]interface {}{"noxu.example.com/finalizer"}, "namespace":"", "clusterName":"", "deletionGracePeriodSeconds":0, "name":"foo123", "selfLink":"/apis/mygroup.example.com/v1alpha1/foo123"}, "num":map[string]interface {}{"num1":9223372036854775807, "num2":1000000}, "apiVersion":"mygroup.example.com/v1alpha1", "content":map[string]interface {}{"key":"value"}, "kind":"WishIHadChosenNoxu"}
        Error Trace:    finalization_test.go:51
        Error:          Received unexpected error "the server reported a conflict (put noxus.mygroup.example.com foo123)"
                
        finalization_test.go:54: get object map[string]interface {}{"apiVersion":"mygroup.example.com/v1alpha1", "content":map[string]interface {}{"key":"value"}, "kind":"WishIHadChosenNoxu", "metadata":map[string]interface {}{"finalizers":[]interface {}{"noxu.example.com/finalizer"}, "namespace":"", "resourceVersion":"5", "selfLink":"/apis/mygroup.example.com/v1alpha1/foo123", "name":"foo123", "uid":"4f094c89-4206-11e7-bb84-80e65026e166", "clusterName":"", "creationTimestamp":"2017-05-26T11:27:32Z", "deletionGracePeriodSeconds":0, "deletionTimestamp":"2017-05-26T11:27:32Z"}, "num":map[string]interface {}{"num1":9223372036854775807, "num2":1000000}}
I0526 21:27:32.142029    6573 customresource_discovery_controller.go:163] Shutting down DiscoveryController
I0526 21:27:32.142055    6573 naming_controller.go:295] Shutting down NamingConditionController
FAIL
FAIL    k8s.io/kubernetes/staging/src/k8s.io/kube-apiextensions-server/test/integration 0.678s
```
```
make test-integration WHAT=./staging/src/k8s.io/kube-apiextensions-server/test/integration/ KUBE_TEST_ARGS="-run TestFinalization"
+++ [0526 21:43:28] Checking etcd is on PATH
/Users/ash2k/gopath/src/k8s.io/kubernetes/third_party/etcd/etcd
+++ [0526 21:43:28] Starting etcd instance
etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/tmp.KCVNUY8D --listen-client-urls http://127.0.0.1:2379 --debug > "/dev/null" 2>/dev/null
Waiting for etcd to come up.
+++ [0526 21:43:28] On try 1, etcd: : http://127.0.0.1:2379
{"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
+++ [0526 21:43:28] Running integration test cases
Running tests for APIVersion: v1
+++ [0526 21:43:34] Running tests without code coverage
W0526 21:43:38.181188   14865 authentication.go:50] Authentication is disabled
I0526 21:43:38.183867   14865 serve.go:85] Serving securely on 127.0.0.1:51007
I0526 21:43:38.183931   14865 crd_finalizer.go:249] Starting CRDFinalizer
I0526 21:43:38.183958   14865 naming_controller.go:284] Starting NamingConditionController
I0526 21:43:38.183970   14865 customresource_discovery_controller.go:152] Starting DiscoveryController
2017-05-26T21:43:38.217871492+10:00 AUDIT: id="4307d333-8d3c-4138-8d06-8ae25c94e7aa" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/healthz" response="200"
2017-05-26T21:43:38.218455383+10:00 AUDIT: id="63023714-83ef-4fee-a4c7-70ebeb2adce8" ip="127.0.0.1" method="get" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1" response="200"
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"apiextensions.k8s.io/v1alpha1","resources":[{"name":"customresourcedefinitions","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["create","delete","deletecollection","get","list","patch","update","watch"]},{"name":"customresourcedefinitions/status","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["update"]}]}
E0526 21:43:38.263184   14865 status.go:62] apiserver received an error that is not an metav1.Status: rpc error: code = 13 desc = transport is closing
2017-05-26T21:43:38.219080519+10:00 AUDIT: id="5235bb05-f440-4b3a-bd93-56efb3b1d75b" ip="127.0.0.1" method="create" user="<none>" groups="<none>" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions" response="500"
--- FAIL: TestFinalization (0.33s)
        Error Trace:    finalization_test.go:24
        Error:          Received unexpected error "an error on the server (\"unknown\") has prevented the request from succeeding (post customresourcedefinitions.apiextensions.k8s.io)"
                
FAIL
FAIL    k8s.io/kubernetes/staging/src/k8s.io/kube-apiextensions-server/test/integration 0.410s
```
```
make test-integration WHAT=./staging/src/k8s.io/kube-apiextensions-server/test/integration/ KUBE_TEST_ARGS="-run TestFinalization"
+++ [0526 21:47:21] Checking etcd is on PATH
/Users/ash2k/gopath/src/k8s.io/kubernetes/third_party/etcd/etcd
+++ [0526 21:47:21] Starting etcd instance
etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/tmp.H0T4XrEG --listen-client-urls http://127.0.0.1:2379 --debug > "/dev/null" 2>/dev/null
Waiting for etcd to come up.
/Users/ash2k/gopath/src/k8s.io/kubernetes/hack/lib/util.sh: line 33: 15573 Segmentation fault: 11  etcd --advertise-client-urls http://${ETCD_HOST}:${ETCD_PORT} --data-dir ${ETCD_DIR} --listen-client-urls http://${ETCD_HOST}:${ETCD_PORT} --debug 2> "${ETCD_LOGFILE}" > /dev/null
!!! [0526 21:47:43] Timed out waiting for etcd:  to answer at http://127.0.0.1:2379/v2/machines; tried 80 waiting 0.25 between each
!!! [0526 21:47:43] Call tree:
!!! [0526 21:47:43]  1: hack/make-rules/test-integration.sh:67 kube::etcd::start(...)
!!! [0526 21:47:43]  2: hack/make-rules/test-integration.sh:98 runTests(...)
+++ [0526 21:47:43] Cleaning up etcd
+++ [0526 21:47:43] Integration test cleanup complete
make: *** [test-integration] Error 1
```
This is obviously not acceptable to merge. I don't think it is a test though. Looks like some server-side issues.

--- a/staging/src/k8s.io/kube-apiextensions-server/Godeps/Godeps.json
+++ b/staging/src/k8s.io/kube-apiextensions-server/Godeps/Godeps.json
@@ -255,6 +255,10 @@
 			"Rev": "a22138067af1c4942683050411a841ade67fe1eb"
 		},
 		{
+			"ImportPath": "github.com/pmezard/go-difflib/difflib",
+			"Rev": "d8ed2627bdf02c080bf22230dbb337003b7aba2d"
+		},
+		{
 			"ImportPath": "github.com/prometheus/client_golang/prometheus",
 			"Rev": "e51041b3fa41cece0dca035740ba6411905be473"
 		},
@@ -281,6 +285,14 @@
 		{
 			"ImportPath": "github.com/spf13/pflag",
 			"Rev": "9ff6c6923cfffbcd502984b8e0c80539a94968b7"
+		},
+		{
+			"ImportPath": "github.com/stretchr/testify/assert",
+			"Rev": "e3a8ff8ce36581f87a15341206f205b1da467059"
+		},
+		{
+			"ImportPath": "github.com/stretchr/testify/require",
+			"Rev": "e3a8ff8ce36581f87a15341206f205b1da467059"
 		},
 		{
 			"ImportPath": "github.com/ugorji/go/codec",

--- a/staging/src/k8s.io/kube-apiextensions-server/test/integration/BUILD
+++ b/staging/src/k8s.io/kube-apiextensions-server/test/integration/BUILD
@@ -11,6 +11,7 @@ go_test(
     name = "go_default_test",
     srcs = [
         "basic_test.go",
+        "finalization_test.go",
         "registration_test.go",
     ],
     tags = [
@@ -19,6 +20,7 @@ go_test(
     ],
     deps = [
         "//vendor/github.com/coreos/etcd/clientv3:go_default_library",
+        "//vendor/github.com/stretchr/testify/require:go_default_library",
         "//vendor/k8s.io/apimachinery/pkg/api/errors:go_default_library",
         "//vendor/k8s.io/apimachinery/pkg/api/meta:go_default_library",
         "//vendor/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library",

--- a/staging/src/k8s.io/kube-apiextensions-server/test/integration/finalization_test.go
+++ b/staging/src/k8s.io/kube-apiextensions-server/test/integration/finalization_test.go
@@ -1,0 +1,98 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiextensionsv1beta1 "k8s.io/kube-apiextensions-server/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/kube-apiextensions-server/test/integration/testserver"
+)
+
+func TestFinalization(t *testing.T) {
+	stopCh, apiExtensionClient, clientPool, err := testserver.StartDefaultServer()
+	require.NoError(t, err)
+	defer close(stopCh)
+
+	noxuDefinition := testserver.NewNoxuCustomResourceDefinition(apiextensionsv1beta1.ClusterScoped)
+	noxuVersionClient, err := testserver.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, clientPool)
+	require.NoError(t, err)
+
+	ns := "not-the-default"
+	name := "foo123"
+	noxuResourceClient := NewNamespacedCustomResourceClient(ns, noxuVersionClient, noxuDefinition)
+
+	instance := testserver.NewNoxuInstance(ns, name)
+	instance.SetFinalizers([]string{"noxu.example.com/finalizer"})
+	createdNoxuInstance, err := instantiateCustomResource(t, instance, noxuResourceClient, noxuDefinition)
+	require.NoError(t, err)
+
+	uid := createdNoxuInstance.GetUID()
+	err = noxuResourceClient.Delete(name, &metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &uid,
+		},
+	})
+	require.NoError(t, err)
+
+	// Deleting something with a finalizer sets deletion timestamp to a not-nil value but does not
+	// remove the object from the API server. Here we read it to confirm this.
+	gottenNoxuInstance, err := noxuResourceClient.Get(name)
+	require.NoError(t, err)
+
+	require.NotNil(t, gottenNoxuInstance.GetDeletionTimestamp())
+
+	// Trying to delete it again to confirm it will not remove the object because finalizer is still there.
+	err = noxuResourceClient.Delete(name, &metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &uid,
+		},
+	})
+	require.NoError(t, err)
+
+	// Removing the finalizers to allow the following delete remove the object.
+	// This step will fail if previous delete wrongly removed the object.
+	for {
+		gottenNoxuInstance.SetFinalizers(nil)
+		_, err = noxuResourceClient.Update(gottenNoxuInstance)
+		if err == nil {
+			break
+		}
+		if !errors.IsConflict(err) {
+			require.NoError(t, err) // Fail on unexpected error
+		}
+		gottenNoxuInstance, err = noxuResourceClient.Get(name)
+		require.NoError(t, err)
+	}
+
+	// Now when finalizer is not there it should be possible to actually remove the object from the server.
+	err = noxuResourceClient.Delete(name, &metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &uid,
+		},
+	})
+	require.NoError(t, err)
+
+	// Check that the object is actually gone.
+	_, err = noxuResourceClient.Get(name)
+	require.Error(t, err)
+	require.True(t, errors.IsNotFound(err), "%#v", err)
+}

